// @segment/analytics-signals-runtime/dist/cjs/index.js
"use strict";
// This file is only for people who plan on installing this package in their npm projects, like us.
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMobileRuntimeCode = exports.MobileSignalsRuntime = exports.MobileRuntimeConstants = exports.Mobile = exports.WebSignalsRuntime = exports.getRuntimeCode = exports.WebRuntimeConstants = exports.SignalsRuntime = void 0;
const tslib_1 = require("tslib");
// shared
var signals_runtime_1 = require("./shared/signals-runtime");
Object.defineProperty(exports, "SignalsRuntime", { enumerable: true, get: function () { return signals_runtime_1.SignalsRuntime; } });
// web
tslib_1.__exportStar(require("./web/web-signals-types"), exports);
tslib_1.__exportStar(require("./shared/shared-types"), exports);
exports.WebRuntimeConstants = tslib_1.__importStar(require("./web/web-constants"));
var get_runtime_code_generated_1 = require("./web/get-runtime-code.generated");
Object.defineProperty(exports, "getRuntimeCode", { enumerable: true, get: function () { return get_runtime_code_generated_1.getRuntimeCode; } });
var web_signals_runtime_1 = require("./web/web-signals-runtime");
Object.defineProperty(exports, "WebSignalsRuntime", { enumerable: true, get: function () { return web_signals_runtime_1.WebSignalsRuntime; } });
// mobile -- we don't need this *yet*, but some day?
exports.Mobile = tslib_1.__importStar(require("./mobile/mobile-signals-types"));
exports.MobileRuntimeConstants = tslib_1.__importStar(require("./mobile/mobile-constants"));
var mobile_signals_runtime_1 = require("./mobile/mobile-signals-runtime");
Object.defineProperty(exports, "MobileSignalsRuntime", { enumerable: true, get: function () { return mobile_signals_runtime_1.MobileSignalsRuntime; } });
var get_runtime_code_generated_2 = require("./mobile/get-runtime-code.generated");
Object.defineProperty(exports, "getMobileRuntimeCode", { enumerable: true, get: function () { return get_runtime_code_generated_2.getRuntimeCode; } });
