// @segment/analytics-signals-runtime/dist/cjs/shared/signals-runtime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignalsRuntime = void 0;
/**
 * Base class that provides runtime utilities for signals.
 */
class SignalsRuntime {
    constructor(signals = []) {
        /**
         * Finds a signal of a specific type from a given signal.
         *
         * SignalType - The type of the signal to find.
         * @param fromSignal - The signal to search from.
         * @param signalType - The type of the signal to find.
         * @param predicate - Optional predicate function to filter the signals.
         * @returns The found signal of the specified type, or undefined if not found.
         */
        this.find = (fromSignal, signalType, predicate) => {
            return this.filter(fromSignal, signalType, predicate)[0];
        };
        /**
         * Filters signals of a specific type from a given signal.
         * SignalType - The type of the signals to filter.
         * @param fromSignal - The signal to search from.
         * @param signalType - The type of the signals to filter.
         * @param predicate - Optional predicate function to filter the signals.
         * @returns An array of signals of the specified type.
         */
        this.filter = (fromSignal, signalType, predicate) => {
            const _isSignalOfType = (signal) => signal.type === signalType;
            return this.signalBuffer
                .slice(this.signalBuffer.indexOf(fromSignal) + 1)
                .filter(_isSignalOfType)
                .filter((signal) => (predicate ? predicate(signal) : () => true));
        };
        // initial signals
        this.signalBuffer = signals;
    }
}
exports.SignalsRuntime = SignalsRuntime;
