// @segment/analytics-signals-runtime@1.5.0/dist/cjs/mobile/mobile-constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalDataAction = exports.NetworkAction = exports.NavigationAction = exports.EventType = exports.SignalType = void 0;
exports.SignalType = Object.freeze({
    Interaction: 'interaction',
    Navigation: 'navigation',
    Network: 'network',
    LocalData: 'localData',
    Instrumentation: 'instrumentation',
    UserDefined: 'userDefined',
});
exports.EventType = Object.freeze({
    Track: 'track',
    Page: 'page',
    Screen: 'screen',
    Identify: 'identify',
    Group: 'group',
    Alias: 'alias',
});
exports.NavigationAction = Object.freeze({
    Forward: 'forward',
    Backward: 'backward',
    Modal: 'modal',
    Entering: 'entering',
    Leaving: 'leaving',
    Page: 'page',
    Popup: 'popup',
});
exports.NetworkAction = Object.freeze({
    Request: 'request',
    Response: 'response',
});
exports.LocalDataAction = Object.freeze({
    Loaded: 'loaded',
    Updated: 'updated',
    Saved: 'saved',
    Deleted: 'deleted',
    Undefined: 'undefined',
});
