// .ignored/@segment/analytics-signals/dist/cjs/types/factories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createNetworkSignal = exports.createUserDefinedSignal = exports.createNavigationSignal = exports.createInteractionSignal = exports.createInstrumentationSignal = void 0;
const normalize_url_1 = require("../lib/normalize-url");
const page_data_1 = require("../lib/page-data");
/**
 * Base Signal Factory
 */
const createBaseSignal = (type, data) => {
    return {
        timestamp: new Date().toISOString(),
        anonymousId: '',
        type,
        data: {
            ...data,
            page: (0, page_data_1.getPageData)(),
        },
    };
};
const createInstrumentationSignal = (rawEvent) => {
    return createBaseSignal('instrumentation', { rawEvent });
};
exports.createInstrumentationSignal = createInstrumentationSignal;
const createInteractionSignal = (data) => {
    return createBaseSignal('interaction', data);
};
exports.createInteractionSignal = createInteractionSignal;
const createNavigationSignal = (data) => {
    return createBaseSignal('navigation', data);
};
exports.createNavigationSignal = createNavigationSignal;
const createUserDefinedSignal = (data) => {
    return createBaseSignal('userDefined', data);
};
exports.createUserDefinedSignal = createUserDefinedSignal;
const createNetworkSignal = (data, metadata) => {
    return {
        ...createBaseSignal('network', {
            ...data,
            url: (0, normalize_url_1.normalizeUrl)(data.url),
        }),
        metadata: metadata ?? {
            filters: {
                allowed: [],
                disallowed: [],
            },
        },
    };
};
exports.createNetworkSignal = createNetworkSignal;
