// .ignored/@segment/analytics-signals/dist/esm/types/factories.js
import { normalizeUrl } from '../lib/normalize-url';
import { getPageData } from '../lib/page-data';
/**
 * Base Signal Factory
 */
const createBaseSignal = (type, data) => {
    return {
        timestamp: new Date().toISOString(),
        anonymousId: '',
        type,
        data: {
            ...data,
            page: getPageData(),
        },
    };
};
export const createInstrumentationSignal = (rawEvent) => {
    return createBaseSignal('instrumentation', { rawEvent });
};
export const createInteractionSignal = (data) => {
    return createBaseSignal('interaction', data);
};
export const createNavigationSignal = (data) => {
    return createBaseSignal('navigation', data);
};
export const createUserDefinedSignal = (data) => {
    return createBaseSignal('userDefined', data);
};
export const createNetworkSignal = (data, metadata) => {
    return {
        ...createBaseSignal('network', {
            ...data,
            url: normalizeUrl(data.url),
        }),
        metadata: metadata ?? {
            filters: {
                allowed: [],
                disallowed: [],
            },
        },
    };
};
